### PR TITLE
task/1593 linked selection histogram

### LINKED ISSUE #1593: scatter plot demo with linked density histograms
References: #1585

For reference, here are feature that would be nice to have eventually:

* on a single canvas (using kiwi layout)
* with `bokeh-server` but without polling
* without the `bokeh-server` (using a function interface)

But for now we can build this demo with simple grid/table layout and routing events through `bokeh-server` using polling (or in the notebook using `push_notebook`)

### PR DESCRIPTION
**Pull request by bryevdv**
issues: closes #1593

**NOTE** requires a BokehJS rebuild (adds a `select_every_mousemove` property to `LassoSelectionTool`) 

@pzwang can you check this out and provide feedback? If necessary, is just the one historgram sufficient (for now)? I don't believe the grid plot layout can accept an "empty" space just yet, so laying out a vertical histogram might be a little tricky (or we can improve the grid layout but I'd like to move on). 

Considering adding control to set the number of bins, any other suggestions?

### CODE REVIEW

**Comment by pzwang** [comment]
Looks nice!  Agree that a slider to set # of bins would be great, and should be simple to add.

Can you change the bar plot so that the overall histogram is shown just as outlined bars (with fillcolor="white"), and then show the histogram of the selection with filled bars but with a somewhat narrower width?

The more I think about it, the more I think it would be really great to have a vertical histogram.  What is the difficulty with implementing that?  Why is it empty to start with?  (After all, the horizontal histogram is not empty...)

**Comment by bryevdv** [comment]
The histogram is not empty. The fourth (corner) cell in the grid is empty (no plot) and there is no way to spell that at the moment.

**Comment by pzwang** [comment]
How hard would it be to implement a Spacer object which doesn't render anything, but has fixed or resizable width in one or both dimensions?

**Comment by hhuuggoo** [comment]
we have a pre-text widget - I believe you could use it and supply no text and get that effect

**Comment by bryevdv** [comment]
This is incredible, how does the timing of desired new features line up so perfectly 

http://stackoverflow.com/questions/27782759/how-to-create-a-grid-of-plots-in-bokeh-w-some-empty-cells

**Comment by bryevdv** [comment]
I will try the `VBox`/`HBox`/`Pre` but a real solution (I think) is to allow `None` inside the grid plot list, e.g. `GridPlot(children=[[a,b], [c, None])`

**Comment by bryevdv** [comment]
Not sure what's up I can't seem to get `HBox` to work at all, though it works fine on other examples.

**Comment by bryevdv** [comment]
@peter if I tinker with the bar width, it's no longer a histogram. I'll find some other way to underlay the original data, though.

**Comment by bryevdv** [comment]
screen shot 2015-01-05 at 10 21 43 am

@pzwang better? I maintain bin sizes from the original binning in all updates now

I am going to push what I have, @mattpap if you can tell me what dumb thing I am doing that breaks `HBox` it would be very appreciated.

**Comment by pzwang** [comment]
> how does the timing of desired new features line up so perfectly

90% of being lucky is making your own luck.

> if I tinker with the bar width, it's no longer a histogram

Well, I was just talking about a visual tweak, not actually re-binning in dataspace. But it occurs to me that the selection histogram bar heights can never exceed the baseline, so it's moot.

Another mode to think about:  Drawing the selection histogram above the x-axis, and the remainder/unselected histogram below the X-axis, on an inverted axis.  (Note: unselected, not baseline.)  So, it's like a horizontal tornado plot.  Would be very pretty.... :-)

**Comment by bryevdv** [comment]
Unfortunately we don't have really capability to have a a single axis that inverts at the origin. You could do twin axes, but then the sizes of the plots would not line up right because the histograms would have more axes than the original plot.

**Comment by bryevdv** [comment]
OK inverted the colors, makes more sense with the selected hist being dark like the selected points (as @pzwang originally suggested) Also here is a screen shot that shows the broken layout:
screen shot 2015-01-05 at 10 54 45 am

And here is the layout code:
```
layout = VBox(HBox(p, pv), HBox(ph, pfill))
show(layout) # open a browser
```

I don't see what's wrong with this. Other examples with both `VBox` and `HBox` work fine. I have also tried revering the nesting (`VBox`es inside an `HBox`) but to no avail.

**Comment by bryevdv** [comment]
While @mattpap looks into the widget layout option, I will look at adding real support in `GridPlot` for `None`

**Comment by bryevdv** [comment]
OK, `GridPlot` layout also fails spectacularly on the server. However, works fine on `output_file` is this some kind of CSS conflict? @mattpap @hhuuggoo

**Comment by tr8dr** [comment]
Yeah, exactly what I am looking for (that was my stackoverflow posting).

**Comment by mattpap** [comment]
This may be related to stateful behavior of plotting API. `{V,H}Box` are not being rendered at all, but plots are. I'm investigating this.

**Comment by bryevdv** [comment]
`VBox`/`HBox` combo also works (mostly) with `output_file` but not on server.

**Comment by bryevdv** [comment]
@tr8dr We should hopefully have a workaround today or failing that something implemented in the next dev or point release.

**Comment by bryevdv** [comment]
@mattpap yah that thought has also occurred to me. The other examples that work are mostly glyph examples that manage their own documents. There is alot less state in `bokeh.plotting` than there used to be, but we obviously have further to go to root it all out if that is feasible.

**Comment by bryevdv** [comment]
@mattpap also to note, `plotting/server/grid.py` works just fine.

**Comment by mattpap** [comment]
I think I found the problem.

**Comment by bryevdv** [comment]
A few notes, there are drawback potential to both approaches:
* with `VBox`/`HBox` an annoying level care is required to align things if, for instance, only one of the plots is desired to have a toolbar
* This is better with `GridPlot`, but currently all subplots share the same tool set which is limiting

**Comment by bryevdv** [comment]
@mattpap should I wait to merge any final improvements? Are you making a change on this branch or another? What is the issue?

**Comment by mattpap** [comment]
@bryevdv, I'm still debugging. Any fixes will be submitted in another PR.

File: examples/plotting/server/selection_update.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+renderer = p.select(dict(name="scatter"))
+scatter_ds = renderer[0].data_source
+
+box_select_tool = p.select(dict(type=BoxSelectTool))
+box_select_tool.select_every_mousemove = False
+lasso_select_tool = p.select(dict(type=LassoSelectTool))
+lasso_select_tool.select_every_mousemove = False
+
+# create the horizontal historgram
```
Comment (damianavila) [inline examples/plotting/server/selection_update.py]: histogram

File: examples/plotting/server/selection_update.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+ph = figure(toolbar_location=None, plot_width=p.plot_width, plot_height=200, x_range=p.x_range, title=None, min_border=10, min_border_left=50)
+ph.quad(bottom=0, left=hedges[:-1], right=hedges[1:], top=hhist, color="white", line_color="#3A5785")
+ph.quad(bottom=0, left=hedges[:-1], right=hedges[1:], top=hzeros, color="#3A5785", alpha=0.6, line_color=None, name="hhist")
+ph.xgrid.grid_line_color = None
+
+renderer = ph.select(dict(name="hhist"))
+ph_source = renderer[0].data_source
+
+# create the vertical historgram
```
Comment (damianavila) [inline examples/plotting/server/selection_update.py]: another typo

File: examples/plotting/server/selection_update.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+vhist, vedges = np.histogram(y, bins=20)
+vzeros = [0]*(len(vedges)-1)
+
+# need to adjust for toolbar height, unfortunately
```
Comment (damianavila) [inline examples/plotting/server/selection_update.py]: yep, that's bummer... but it is ok for now...

**Comment by damianavila** [comment]
> @damianavila would like to merge #1623 into master, as well as @mattpap's layout fix, then merge master into this PR. I have no other changes planned however.

OK, sounds like a good plan...

**Comment by mattpap** [comment]
@bryevdv, after all I "fixed" the issue in-place. Please try it out.

In general, it's not a complete fix, because complete fix asks for substantial redesign of plotting API. Currently, when one issues `show(p)`, then he could write `show()` as well. There is no difference in the server context. Grid plots work, because there is `gridplot()` function that handles this case (manually removes figures from `PlotContext`). I implemented `hbox()` and `vbox()` functions that do the same thing, so that we have something working right now, but if someone uses `HBox` or `VBox` directly, he will encounter the original bug.

**Comment by bryevdv** [comment]
I am pretty -1 on adding `hbox` and `vbox` just to make this work in the short term, because I'd expect them to otherwise be rendered unnecessary in the near future, and I don't want to have to go through a deprecation cycle with them.  As a compromise I would suggest this: add the private functions into `bokeh.plotting` but add `hbox` and `vbox` directly to the example (and @tr8dr could use them as a workaround for now, too). 

You're right the best solution is to continue the removal of state from `bokeh.plotting`. I'd like your thoughts on this in a separate issue. Perhaps we can have a `bokeh.plotting.init(...)` that creates a `Document` and a `Session` (as necessary). Or maybe `output_thing(...)` can return a handle to something that gets used explicitly later.

**Comment by bryevdv** [comment]
@pzwang Here is current output, think the features, etc. are done. We just need to resolve the issue described above. 

screen shot 2015-01-05 at 5 20 09 pm

I am planning to make some YT videos to describe this example and some others in more detail.

**Comment by pzwang** [comment]
BEAUTIFUL.

**Comment by bryevdv** [comment]
`GridPlot` with `None` to be handled in #1636

**Comment by bryevdv** [comment]
@tr8dr we are going to clean up this particular use-case more for 0.8 because it will require fixing some underlying problems in `bokeh.plotting`. But you can use the code from this PR as a short-term workaround to get you going. Hopefully it is pretty clear which small parts you want, basically: `_deduplicate_plots`, `_push_or_save` and `hbox` and `vbox`. However please let us know if you have any questions or run into trouble.

**Comment by bryevdv** [comment]
@damianavila I plan to merge this as soon as tests pass

**Comment by damianavila** [comment]
That's OK... last changes LGTM.

**Pull request merged**